Switch all remaining targets to ESM-only

devversion opened the pull request:
See individual commits

devversion committed:
refactor: update packages/common tests to work with ESM pipeline

We no longer have generated `.js` files as everything generated
is the ESM `.mjs` output. The tests need to be updated to reflect that.

devversion committed:
fixup! build: update dev-infra build tooling to leverage recent ESM changes

ds

devversion committed:
build: update `packages/bazel/test` to work with ESM

The bazel tests cannot use CommonJS features like `require` and should
be updated to work with ESM. Additionally some specs are updated to
no longer assert CJS/UMD output from `ng_module`. The rule only emits
ESM now.

devversion committed:
build: refactor `packages/compiler/test` to work with ESM

devversion committed:
refactor: update compiler-cli babel linker to be ESM only

Since the linker is no longer being tested with CommonJS, we can remove
most of the CJS/ESM interop trickery.

devversion committed:
refactor: change ngcc to only rely on ESM features

Refactors ngcc to only rely on ESM features because CJS
is no longer needed & tested.

devversion committed:
refactor: update compiler-cli to work with ESM

Updates compiler-cli & tests to be full ESM compatible. Tests
no longer with CommonJS.

devversion committed:
fixup! build: bundle tests for karma web test suites

ds

devversion committed:
refactor: update platform-browser tests to run as ESM

`platform-browser` tests now run in ESM and with `.mjs` output, so
the build targets and tests need to be updated.

Here we change the `zone_event_unpatched` script to include the
`.init` suffix that will be picked up by `spec_bundle`.

Also some circular dependency tests are updated to refer to the
`.mjs` files.

devversion committed:
refactor: update image-directive e2e test to work with ESM

* Switches to the canonical dev-infra http server
* Uses the bundle for serving.
* Switches app_bundle to simple `esbuild` since the test relies
  on `ngDevMode` which `app_bundle` elides as optimization.

devversion committed:
build: introduce `zone_compatible_jasmine_node_test` rule

Introduces a variant of `jasmine_node_test` that works with async/await
in combination with ZoneJS. This is needed for tests relying on
async/await in combinatin with change detection.

Browser tests are already benefting from ESbuild async-await
downleveling. Node tests would ideally not need this, but until
ZoneJS supports native async/await we need to also process
source files to avoid native async/await syntax.

devversion committed:
refactor: update `packages/router` tests to be compatible with ESM

The ESM js files need to be referenced, and the router tests rely
on `async/await` with change detection- so a special rule is required
that downlevels `async/await` to generators (similar to the Angular CLI)

devversion committed:
build: add missing dependencies to platform-server and tools/testing

There are two build targets which never had all its runtime dependencies
properly specified. This wasn't noticed because there were macros in
`defaults.bzl` that automatically included these deps.

In a follow-up we will clean-up this legacy auto-deps feature in
`defaults.bzl`.

devversion committed:
build: bundle tests for protractor tests to support ESM output

Protractor does not support ESM, so we need to take all the ESM
output and bundle it into a CommonJS file. This comes at the cost
of not using actual ESM for execution, but the ESBuild bundle follows
strict ESM semantics so we can be sure it's compatible when we have
an ESM-compatible e2e test runner in the future.

devversion committed:
refactor: update `packages/localize` to be strict ESM

The `packages/localize` package still required some trickery
to support CommonJS because tests in the repo were running as CommonJS.

This commit removes the CommonJS logic in localize and its tests, so
that only ESM is used in production & tests.

devversion committed:
refactor: update `platform-server` tests to work with ESM

* Uses `.mjs` for circular deps tests
* Replaces `require` with ESM-equivalent. uses an import statement here
  for proper typing, and specify dependency properly.

devversion committed:
refactor: update `platform-browser-dynamic` to work with ESM output

* Switches circular dependency tests to use the `.mjs` output.

devversion committed:
refactor: update `upgrade` package tests to work with ESM output

* Uses the `.mjs` output of the upgrade package when checking for
  circular deps.

devversion committed:
refactor: update `forms` package tests to work with ES2020 ESM

Tests now always run with ESM 2020, while previously they ran with
ES2015 CommonJS UMD bundles.

Since ZoneJS does not support intercepting native `async/await` syntax,
the forms test needs to use the zone-compatible variant of
`jasmine_node_tests`. This variant downlevels the native `async/await`
syntax to generators that ZoneJS can intercept. All of this is done
using the dev-infra ESBuild `spec_bundle` rule.

devversion committed:
refactor: update `elements` test to work with ESM outputs

* Uses the `.mjs` ESM output for testing/checking circular dependencies.

devversion committed:
build: ensure esbuild `nodejs_binary` correctly wires up ESM loader

ESBuild relies on the linker and we currently set up the ESM loader,
along with accidentally enabling the patched resolution loader. This
didn't cause any problems in sandbox, but outside of sandbox incorrect
ESBuild versions may be discovered because the loader looks at the
top-level `npm/` node modules before looking relative to e.g.
`@bazel/esbuild`

devversion committed:
build: always attempt next/builtin resolution first when resolving modules

Even with patched resolution, we should always attempt the next/builtin
NodeJS resolution first. It may find a module if there `node_modules`
relative to the context file. This would be more correct than looking
for a module always at the Bazel `npm` repository `node_modules` folder.

devversion committed:
refactor: convert AIO tooling scripts used in Bazel to ESM

Since the Bazel setup in this repo will now always use ESM,
the tooling scripts/binaries in AIO need to be switched to ESM
too. Most of the scripts are already ESM, but a few had to be converted.

Note that the Dgeni generation does not use ESM because it's unaffected
and the Dgeni CLI is used. In the future we could also update the Dgeni
setup to ESM but there is no need currently.

devversion committed:
build: update dev-infra build tooling and properly handle builtin `node:` imports

* Updates build-tooling to benefit from the latest `spec_bundle`
  improvements.
* Updates the ESM extension loader to not attempt adding extensions to
  builtin `node:` specifiers. This seems to be disallowed and cannot be
  handled gracefully (the attempts are part of a try/catch).

```

Use --sandbox_debug to see verbose messages from the sandbox
Error [ERR_UNKNOWN_BUILTIN_MODULE]: No such built-in module: node:fs.mjs
    at new NodeError (node:internal/errors:371:5)
    at ESMLoader.builtinStrategy (node:internal/modules/esm/translators:276:11)
    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:236:14)
```

devversion committed:
refactor: simplify devtools setup given recent ESM changes

Since the `defaults.bzl` repo-wide macros are now supporting ESM,
the special spec-bundle logic from `devtools` can be removed.

Also the esbuild configurations need to be updated to account
for the recent dev-infra build-tooling changes. Also properly
now ensures that `aysnc/await` is downleveled for ZoneJS compatibility.

devversion committed:
refactor: switch `packages/examples` to ESM-compatible http server

* Switches all examples to use dev-infra's canonical ESM-compatible
  `http_server`.
* Uses ESBuild for bundling ESM into a single file, compared to having
  to load hundreds of individual ESM files in the browser (potential
  source of flakiness & slowness).

devversion committed:
refactor: update `tools` tests and scripts to work with Bazel ESM

* The Karma Bazel Saucelabs binary needs to use `.mjs` as everything in
  the repo w/Bazel is supposed to be ESM.
* The symbol extractor test is updated to no longer use CommonJS
  features like `require.resolve`.

devversion committed:
refactor: update `language-service` package and tests to work with ESM

The Angular Language Service package is now tested & compiled using ESM.
Previously it was a mismatch of CommonJS and ESM.

Still the language-service ESM output is transformed into a single UMD
bundle to allow for module resolution to be overriden. See
`bundles/BUILD`.

This is kept as is. To fully ship ESM (language-service is an exception
here), we need to:

 * Update all code to no longer reference typescript via import. Instead
   typescript needs to be passed around so that the extension can
   control the version
 * The VSCode extension/ the TS server needs to be able to load ESM. It
   looks like the server supports ESM global plugins, but the extension
   might not yet.

This is out of scope for the dev-infra effort as it requires more
insight into VSCode & the extension system & the TS language server.

devversion committed:
refactor: update `service-worker` tests to work with ESM

* Updates circular dependency tests to use the `.mjs` outputs
* Switches away from CommonJS specific `require` calls.
* Simplifies the test helper logic since all browsers/NodeJS versions
  support `URL` as a global.

devversion committed:
refactor: update `modules/` benchmarks and playgrounds to work with ESM

* Switches away from the ESM-incompatible & unmaintained `ts_devserver`
  to `http_server`. This is the canonical server maintained by dev-infra
* Switches tests away from CommonJS specific logic. e.g. require.resolve
* Adjusts tests to work with Protractor spec bundling (Protractor does
  not support ESM execution, but we want to take ESM-written specs)
* Reworks playground and benchmarks to use `app_bundle` and `esbuild`
  instead of loading hundreds of files individually. This also makes
  tests more stable and more aligned with real applications.

devversion committed:
build: update build-tooling and adjust benchmark macros for ESM

* The benchmark macro should also use devmode ESM 2020. No CommonJS
* The benchmark macro should always add `benchpress` as runtime
  dependency because it is loaded asynchronously.
* The protractor `nodejs_binary` should use our ESM-interop binary
so that ESM resolution works (e.g. when `await import(benchpress)` from
the driver utilities is invoked).

devversion committed:
remove prodmode language-service cjs

devversion committed:
refactor: update `animations` tests to work with ESM output

The circular deps tests should use the `.mjs` output.

devversion committed:
refactor: update `zone.js` and tests to work with ESM

* Adjusts tests to no longer rely on CommonJS features. Switches them to
  ESM
* Updates test initialization files to not double-initialize Jasmine now
  that bootstrap files are loaded after Jasmine. The `jasmine.boot`
  setup was hacky from `rules_nodejs` and will break in the future
  regardless if we e.g. use `rules_js` with actual unmodified `jasmine`.

devversion committed:
fixup! build: refactor `ng_package` to work with ESM

ds

devversion committed:
fixup! build: refactor `generate-locales-tool` to ESM

ds

devversion committed:
fixup! test: refactor render3 definition feature test to be less sensitive to JS output

ds

devversion committed:
fixup! build: support enabling esm interop loader even outside of `defaults.bzl`

ds

devversion committed:
fixup! refactor: remove `__ESM_IMPORT_META_URL__` workaround now that we can use ESM

ds

josephperrott commented on packages/zone.js/test/node_error_disable_policy_entry_point.init.ts:
```diff
@@ -6,6 +6,5 @@
-process.env['errorpolicy'] = (global as any)['__Zone_Error_ZoneJsInternalStackFrames_policy'] =
-    'disable';
-import './node_error_entry_point';
+import './node_error_disable_policy';
```
We should include a note here at this is import order sensitive.

devversion replied:
done!

josephperrott reviewed (approved):
LGTM

devversion commented on packages/compiler-cli/esbuild.config.js:
```diff
@@ -7,12 +7,12 @@
 
 module.exports = {
-  resolveExtensions: ['.mjs', '.js'],
+  resolveExtensions : ['.mjs', '.js'],
```
yeah, thought too, but clang-format enforces it for some reason

devversion commented on packages/compiler-cli/ngcc/src/execution/cluster/master.ts:
```diff
@@ -343,13 +343,15 @@ export class ClusterMaster {
-      requireFn.resolve('@angular/compiler-cli/ngcc/src/execution/cluster/ngcc_cluster_worker');
-  return fileSystem.resolve(workerScriptPath);
+/** Wrapper for resolving the cluster worker script. Useful for test patching. */
+export class ClusterWorkerScriptResolver {
```
In ESM module members cannot be spied/patched. So for tests to be able to mock, we wrap it in a container. Did the same in dev-infra for e.g. Log

josephperrott replied:
Ahh makes sense

devversion commented on packages/service-worker/worker/testing/utils.ts:
```diff
@@ -55,9 +43,7 @@ export function normalizeUrl(url: string, relativeTo: string): NormalizedUrl {
-  const parsedUrl: URL = (typeof URL === 'function') ?
-      (!relativeTo ? new URL(url) : new URL(url, relativeTo)) :
-      require('url').parse(require('url').resolve(relativeTo || '', url));
+  const parsedUrl: URL = (!relativeTo ? new URL(url) : new URL(url, relativeTo));
```
Potentially. I will keep as is because that's the original code, but agreed that this might be a good cleanup

devversion committed:
fixup! build: create NodeJS ESM loader for supporting Bazel setup

ds

devversion committed:
build: remove components repo unit tests job

After discussion initiated in the framework team (by kkostadinov),
the team has decided to not keeping the `components-repo-unit-tests`
job. This commit removes it.

devversion commented on modules/benchmarks/src/tree/ng2_static/index.ts:
```diff
@@ -31,7 +33,7 @@ function noop() {}
 function init() {
   initTreeUtils();
-  enableProdMode();
+  // enableProdMode();
```
good catch. reverted!

devversion committed:
build: fix legacy saucelabs test bundle generation after ESM changes

ZoneJS is no longer loaded as an UMD, but instead is included as part
of the browser init entry-point. This means that ZoneJS is bundled and
the ESBuild logic needs to be adjusted for that.

devversion committed:
build: remove remaining usages of deprecated `ts_devserver`

* Switches all remaining targets (even if not tested and failing as per
  build) away from `ts_devserver` to the canonical `http_server` from
  dev-infra.
